### DISCUSSION TITLE
How to determine a polygon_with_holes is simple.

### DISCUSSION THREAD

**Discussion by ChenKe404** [CGAL/cgal#8124] [Q&A] [answered]
## Issue Details

I want to determine a Polygon_with_holes is simple or not, i only found "CGAL::is_simple_2", but this function just only handle Polygon_2.
![image](https://github.com/CGAL/cgal/assets/75658944/3b0453db-25af-4c72-a469-c6e6b3ec73b3)
of these polygons, I think only numbered 5 and 3 are simple, but actually 5 always not simpale.

## Source Code

```c++
bool isSimple(const Polygon_with_holes_2& pgn)
{
    // outer is simple ?
    auto& outer = pgn.outer_boundary();
    if(!outer.is_simple())
        return false;
    auto& holes = pgn.holes();
    
    // holes must simple and can't intersect or touch with outer
    for(auto& h : holes)
    {
        if(!h.is_simple())
            return false;
        // find out hole's points is on outer's bound or out of bound. if not,so this polygon is not simple
        for(auto& p : h)
        {
            if(CGAL::bounded_side_2(outer.begin(),outer.end(),p) != CGAL::ON_BOUNDED_SIDE)
                return false;
        }
    }
    
    // the conditions "hole with hole" is like "hole with outer".
   // I need to find out two holes touch with same point(because polygon 5), but I don't know how to do that.
    for(int i=0;i<holes.size();++i)
    {
        auto& h0 = holes[i];
        for(int j=i+1;j<holes.size();++j)
        {
            auto& h1 = holes[j];
            for(auto& p : h0)
            {
                // if h0 touch h1 at a same point, this place will be return, i don't want this
                if(CGAL::bounded_side_2(h1.begin(),h1.end(),p) != CGAL::ON_UNBOUNDED_SIDE)
                    return false;
            }
            for(auto& p : h1)
            {
                if(CGAL::bounded_side_2(h0.begin(),h0.end(),p) != CGAL::ON_UNBOUNDED_SIDE)
                    return false;
            }
        }
    }
    return true;
}
```
## Environment

* Operating system (Windows10, 32/64 bits):
* Compiler: MSVC2019
* Release or debug mode: debug
* Specific flags used (if any):
* CGAL version: 5.6.1
* Boost version: 1.83
* Other libraries versions if used (Eigen, TBB, etc.):

Sorry about my english.

PS: after that, i need to conver complex polygon to simple, but i don't know how to do this too.

**Comment by lryan599** [comment, marked as answer]
There are some free functions for validating a polygon (or a polygon with hole) in "CGAL/Boolean_set_operations_2/Gps_polygon_validation.h". Perhaps you can refer to it.
Also see https://stackoverflow.com/questions/64468143 and user manual "2D Regularized Boolean Set-Operations" chapter for more detailed information.

**Comment by ChenKe404** [comment]
Thank you. I'll try @lryan599
